arch/lkl/kernel/signal.c

```diff
@@ -28,47 +31,249 @@ static void initialize_uctx(struct ucontext *uctx, const struct pt_regs *regs)
     return;
 }
 
-/* Function to invoke the signal handler of LKL application. This works for 
+/* 
+ * Function to invoke the signal handler of LKL application. This works for 
  * sig handler installed using sigaction or signal API. This will remove the
  * overhead of injecting the stack frame to pass the user context to user
  * space application (could lead to inclusion of ARCH specific code)
+ * 
+ * MUST be called owning the lock.
  */
 static void handle_signal(struct ksignal *ksig, struct ucontext *uctx)
 {
+    lkl_thread_t self;
+
+    /* In DEBUG build check that the lock is currently held by this task */
+    lkl_assert_cpu_owner();
+
+	/*
+	 *  Give up the cpu lock while we invoke the signal handler because we are returning
+     *  to userspace.
+	 */
+
+    lkl_cpu_put();
+
+    /* In case it was recursively locked */
+    lkl_assert_cpu_not_owner();
+
+    /* Get the current thread before we invoke the signal handler */
+    self = lkl_ops->thread_self();
+
     ksig->ka.sa.sa_handler(ksig->sig, (void*)&ksig->info, (void*)uctx);
+
+    /* 
+     *  Check if the apparent current thread changes while processing the signal.
+     *  Should not happen.
+     */
+
+	BUG_ON(!lkl_ops->thread_equal(self, lkl_ops->thread_self()));
+
+    /* take back the lock */
+	lkl_cpu_get();
 }
 
+/*
+ *  Return only the signal of a given number, while snipping it from the list
+ *  Used by the trap handler for sync signals such as SEGV, FPE.
+ */
+int get_given_ksignal(struct thread_info *task, struct ksignal* sig, int which)
+{
+    struct ksignal_list_node *next;
+    struct ksignal_list_node *node;
+    struct ksignal_list_node *previous;
+
+    spin_lock(&task->signal_list_lock);
+    node = task->signal_list_head;
+    previous = NULL;
+
+    while (node != NULL) {
+        next = node->next;
+        if (node->sig.sig == which) { 
+           /* first remove the node we found from the linked list */
+           
+            if (previous != NULL) /* not at the front? */
+                previous->next = next;
+            else
+                task->signal_list_head = next;
+
+            if (next == NULL)                       /* was that the last node? */
+                task->signal_list_tail = previous;  /* yes, have the tail point at the previous or be null */
+
+            spin_unlock(&task->signal_list_lock);
+            memcpy(sig, &node->sig, sizeof(*sig));  /* copy the signal back to the caller */
+            LKL_TRACE("Fetching task %p signal %d\n", current, sig->sig);
+            kfree(node);
+            return 1;
+        }
+        previous = node;
+        node = next;
+    }
+    
+    spin_unlock(&task->signal_list_lock);
+    return 0;
+}  
+
+int get_next_ksignal(struct thread_info *task, struct ksignal* sig)
```

Comment:
Maybe add comment that "return 1" means there is a signal, "return 0" means no signal.

Reply:
ok